### PR TITLE
Feature/ACL

### LINKED ISSUE #225: [request] ACL
Currently, we can create a list of APIs and a list of Consumers. Any consumer has access to any API. Instead we would like to create a to be API accessible only to a set of consumers. Something like a Premium plan to access a API. Do we have something already available? Or some way to achieve this? 

We can do this adding the list of consumers who have access to the API in the /apis resource. 
Or else, have a separate private-apis plugin and maintain the mapping of APIs and Consumers in a separate column family. 

Please let us know your suggestions. We would like send a PR for the change.

### PR DESCRIPTION
**Pull request by thefosk**
Closes #225. Implements a basic ACL control that allows to whitelist and blacklist consumers to APIs using arbitrary group names. The ACL plugin could be more powerful once #505 has been implemented.

The documentation on the website has been updated, but it needs an ACL icon (@jitachigarcia).

### CODE REVIEW

File: kong/plugins/acl/daos.lua [hunk 1]
```diff
@@ -0,0 +1,22 @@
+local BaseDao = require "kong.dao.cassandra.base_dao"
+
+local SCHEMA = {
+  primary_key = {"id"},
+  fields = {
+    id = { type = "id", dao_insert_value = true },
+    created_at = { type = "timestamp", dao_insert_value = true },
+    consumer_id = { type = "id", required = true, foreign = "consumers:id" },
```
Comment (thibaultCha) [inline kong/plugins/acl/daos.lua]: If foreign, the field also needs to be `queryable = true` and have an index.

**Comment by DanielDePauw** [comment]
Would it be possible to also control a specific request type? 
Example: consumer1 could only POST to API1 but could GET from API2?

The same effects could be achieved by expanding the rate limiting plugin to allow you to set limits for consumers using specific methods (GET, POST, etc) to specific APIs. 
Example: 
consumer1 has a POST limit to API1 of 999999 but a GET limit of 0 to API1
consumer1 also has a POST limit of 0 to API2 and a GET limit of 999999 to API2

**Comment by thefosk** [comment]
@DanielDePauw it seems like that belongs to issue #505, so that we could have HTTP method filtering for every plugin (including ACL and Rate Limiting).

File: kong/plugins/acl/daos.lua [hunk 2]
```diff
@@ -0,0 +1,22 @@
+local BaseDao = require "kong.dao.cassandra.base_dao"
+
+local SCHEMA = {
+  primary_key = {"id"},
+  fields = {
+    id = { type = "id", dao_insert_value = true },
+    created_at = { type = "timestamp", dao_insert_value = true },
+    consumer_id = { type = "id", required = true, foreign = "consumers:id", queryable = true },
```
Comment (thefosk) [inline kong/plugins/acl/daos.lua]: @thibaultCha added `queryable`

File: spec/plugins/keyauth/daos_spec.lua [hunk 3]
```diff
@@ -13,7 +13,7 @@ describe("DAO keyauth Credentials", function()
 
   it("should not insert in DB if consumer does not exist", function()
     -- Without a consumer_id, it's a schema error
-    local app_t = {name = "keyauth", value = {key_names = {"apikey"}}}
+    local app_t = {name = "acl", value = {group = {"apikey"}}}
```
Comment (thibaultCha) [inline spec/plugins/keyauth/daos_spec.lua]: ?

**Comment by lucamaraschi** [comment]
Any news on the progress of the PR?

**Comment by thefosk** [comment]
It's done and can be merged.

**Comment by lucamaraschi** [comment]
Super @thefosk! 
Any ETA for the release into master?

**Comment by thefosk** [comment]
@lucamaraschi merged now!

**Pull request merged**